library/std/src/io/buffered/bufwriter.rs

```diff
@@ -291,55 +291,94 @@ impl<W: Write> BufWriter<W> {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<W: Write> Write for BufWriter<W> {
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        // We assume that callers of `write` prefer to avoid split writes where
+        // possible, so we pre-flush the buffer rather than doing a partial
+        // write to fill it.
         if self.buf.len() + buf.len() > self.buf.capacity() {
             self.flush_buf()?;
         }
-        // FIXME: Why no len > capacity? Why not buffer len == capacity? #72919
         if buf.len() >= self.buf.capacity() {
-            self.panicked = true;
-            let r = self.get_mut().write(buf);
-            self.panicked = false;
-            r
+            self.get_mut().write(buf)
         } else {
             self.buf.extend_from_slice(buf);
             Ok(buf.len())
         }
     }
 
     fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
-        // Normally, `write_all` just calls `write` in a loop. We can do better
-        // by calling `self.get_mut().write_all()` directly, which avoids
-        // round trips through the buffer in the event of a series of partial
-        // writes in some circumstances.
-        if self.buf.len() + buf.len() > self.buf.capacity() {
+        // Unlike with `write`, we assume that a caller of `write_all` is
+        // interested in minimizing system calls even if the buffer is split.
+        // This method tries to fill up the buffer as much as possible before
+        // flushing, whereas `write` prefers not split incoming bufs.
+
+        // Bypass the buffer if the the incoming write is larger than the whole
+        // buffer.
+        if buf.len() >= self.capacity() {
             self.flush_buf()?;
+            return self.get_mut().write_all(buf);
         }
-        // FIXME: Why no len > capacity? Why not buffer len == capacity? #72919
-        if buf.len() >= self.buf.capacity() {
-            self.panicked = true;
-            let r = self.get_mut().write_all(buf);
-            self.panicked = false;
-            r
-        } else {
-            self.buf.extend_from_slice(buf);
-            Ok(())
+
+        // In order to reduce net writes in aggregate, we buffer as much as
+        // possible, then forward, then buffer the rest
+        let amt_buffered = self.write_to_buf(buf);
+        if amt_buffered < buf.len() {
+            self.flush_buf()?;
+            // At this point, because we know that buf.len() < self.buf.len(),
+            // we know that this will succeed in totality
+            self.buf.extend_from_slice(&buf[amt_buffered..]);
         }
+        Ok(())
```

Comment:
Lgtm.

This only changes behaviour in the case the `buf` is smaller than the total capacity, but bigger than the available space in the buffer. Before: Flush the buffer, then add all of `buf` to the buffer. After: Add as much of `buf` to the buffer as fits before flushing, then put the remainder of `buf` in the buffer.

Makes sense to me. All of `buf` gets copied anyway and this still results in a single `write` call, but this might reduce the overall number of `write` calls.